src/TabNine.ts

```diff
@@ -1,300 +1,124 @@
 
-        unregisterFunctions.push(() =>
-          this.rl.removeListener("line", onResponse)
-        );
-        this.proc.stdin.write(request, "utf8");
-      } catch (e) {
-        console.log(`Error interacting with TabNine: ${e}`);
-        reject(e);
+        throw new Error("TabNine process is dead.");
       }
-    });
 
-    let timer = new Promise((_resolve, reject) => {
-      let timer = setTimeout(() => reject("request timed out"), timeout);
-
-      unregisterFunctions.push(() => clearTimeout(timer));
-    });
-
-    let procExit = new Promise((_resolve, reject) => {
-      const onClose = () => reject("Child process exited");
-      this.proc.once("exit", onClose);
+      this.proc.stdin.write(
+        JSON.stringify({
+          version: API_VERSION,
+          request: request,
+        }) + "\n",
+        "utf8"
+      );
 
-      unregisterFunctions.push(() => this.proc.removeListener("exit", onClose));
-    });
+      const result = await this.readLineWithLimit(timeout);
 
-    const unregister = () => {
-      unregisterFunctions.forEach((f) => f());
-    };
+      console.log("Received result from binary: ", result);
+      this.consecutiveRestarts = 0;
 
-    return Promise.race([response, timer, procExit]).then(
-      (value) => {
-        unregister();
-        return value;
-      },
-      (err) => {
-        unregister();
-        throw err;
-      }
-    );
+      return JSON.parse(result.toString());
+    } finally {
+      console.log("Binary request failed.");
+      release();
+    }
   }
 
-  private isChildAlive(): boolean {
-    return this.proc && !this.childDead;
-  }
+  private readLineWithLimit(timeout: number): Promise<any> {
+    return new Promise<any>((resolve, reject) => {
+      setTimeout(() => {
+        reject("Binary request timed out.");
+      }, timeout);
 
-  private static runTabNine(
-    context: TabNineExtensionContext,
-    additionalArgs: string[] = [],
-    inheritStdio: boolean = false
-  ): child_process.ChildProcess {
-    const args = [
-      "--client=vscode",
-      "--no-lsp=true",
-      context?.logFilePath ? `--log-file-path=${context.logFilePath}` : null,
-      "--client-metadata",
-      `clientVersion=${context?.vscodeVersion}`,
-      `pluginVersion=${context?.version}`,
-      `t9-vscode-AutoImportEnabled=${context?.isTabNineAutoImportEnabled}`,
-      `t9-vscode-TSAutoImportEnabled=${context?.isTypeScriptAutoImports}`,
-      `t9-vscode-JSAutoImportEnabled=${context?.isJavaScriptAutoImports}`,
-      `vscode-remote=${context?.isRemote}`,
-      `vscode-remote-name=${context?.remoteName}`,
-      `vscode-extension-kind=${context?.extensionKind}`,
-      ...additionalArgs,
-    ].filter(Boolean);
-    const binary_root = path.join(__dirname, "..", "binaries");
-    const command = TabNine.getBinaryPath(binary_root);
-    return child_process.spawn(command, args, {
-      stdio: inheritStdio ? "inherit" : "pipe",
+      this.rl.once("line", resolve);
```

Comment:
This is the thing: there is no need of unregistering the listeners. If the promise resolves with a timeout, the `rl.once(line, resolve)` will just resolve without doing anything (because the promise is already fulfilled).

Reply:
yes, this is correct. however, the concern here is not with promise suddenly being resolved after it was rejected. but that the EventEmitter listener array will endlessly grow, causing a memory leak. 
see the [implementation](https://github.com/nodejs/node-v0.x-archive/blob/ed0d1c384cd4578f7168633c838f1252bddb260e/lib/events.js#L190-L209) of once, where the listener is cleaned only when `on` is invoked

Reply:
It doesn't matter as the GC takes care of that. It will eventually either resolve or the process will die, so we will create a new process and free the pointer to the old process. Then the GC should take care of the rest...